demos/test/force_optimisation.py

```diff
@@ -0,0 +1,1214 @@
+		self.finger_tip_ids = [self.robot_model.getFrameId(name) for name in self.finger_tip_links]
+
+
+
+	def check_values_from_camera(self):
+		if cameraQueue.empty():
+			return
+
+		state_from_camera = cameraQueue.get() # gives array(or matrix) 13x1
+		rotation_matrix = pinocchio.XYZQUATToSe3(state_from_camera).rotation
+
+		diff = pinocchio.difference(self.cube_q_last, state_from_camera)
+
+		
+		finger_relative_position = [self.robot_data.oMf[tip_id].translation - state_from_camera[:3] for tip_id in self.finger_tip_ids]
+		finger_offset_local = [rotation_matrix.T.dot(i) for i in finger_relative_position] # this is the offset (local frame)
+
+		# subtract the offset (rotated in world) from th ewrld tip position
+		# find the avg from the three tips
+
+		'''Now wherever we have the tip positions, we get com as follows
+
+		'''
+		finger_abs_position = [self.robot_data.oMf[tip_id].translation for tip_id in self.finger_tip_ids]
+		new_quat = diff * old_quat
+		self.new_rotation_matrix = pinocchio.Quaternion(new_quat).toRotationMatrix() # this is the new rotation matrix
+		finger_offset_world = [self.new_rotation_matrix.dot(i) for i finger_offset_local] # this rotation matrix is the fixed one
+
+		com = np.mean(np.subtract(finger_abs_position, finger_offset_world)) # this is the new com
+
+		new_cube_state = np.concatenate([com,new_quat]),reshape(-1,1)
+
+		
+
+
+	def box_state_from_tip_position(self):
+		end_effector_position = [self.robot_data.oMf[tip_id].translation for tip_id in self.finger_tip_ids]
+		box_com_position = [(end_effector_position[0][i] + 
+								end_effector_position[1][i] + 
+								end_effector_position[2][i])/3.0 for i in range(3)]
+
+
+		box_com_position = np.resize(box_com_position, (3,1))
+
+		dir_x = np.resize((end_effector_position[1] - end_effector_position[2]),3)
+		dir_x = dir_x / max(np.linalg.norm(dir_x),0.00000001)
+
+		a = np.resize((end_effector_position[2] - end_effector_position[0]),3)
+		a = a /  max(np.linalg.norm(a), 0.000000001)
+
+		b = np.resize((end_effector_position[1] - end_effector_position[0]),3)
+		b = b / max(np.linalg.norm(b), 0.000000001)
+
+		dir_z = np.cross(a,b)
+		dir_z = dir_z / max(np.linalg.norm(dir_z), 0.0000000001)
+
+		dir_y = np.cross(dir_z, dir_x)
+		dir_y = dir_y / max(np.linalg.norm(dir_y), 0.000000001)
+
+		dir_x = dir_x
+		dir_y = dir_y
+		dir_z = dir_z
+
+		self.rotation_matrix = np.matrix([dir_x, dir_y, dir_z]).T
+		self.quat = pinocchio.Quaternion(self.rotation_matrix).coeffs()
+
+		# Setting the position offset
+		offset =  np.array([0,-0.000, 0]).reshape(-1,1)
+
+		box_com_position = box_com_position + self.rotation_matrix.dot(offset)
+
+		# if self.t % 100 == 0:
+		# 	print("Block Position:", box_com_position)
+
+		# self.quat[:] = 0
+		# self.quat[-1] = 1
+		# ipdb.set_trace()
+		# print("Box State:", np.concatenate([box_com_position, self.quat]))
+
+		return np.concatenate([box_com_position, self.quat])
+
+
+		ipdb.set_trace()
```

Comment:
This line should also be removed (it is anyway never reached due to the return above).